### ISSUE TITLE
Hexagon - halide_remote_load_library(libhalide_shared_runtime.so) sysmalloc assertion Aborted

### ISSUE THREAD

**Issue by andrewjong** [halide/Halide#4655]
I am following Halide's Readme [section for Hexagon](https://github.com/halide/Halide#3-build-and-run-an-example-for-hexagon-hvx), trying to run on the simulator.

Using Hexagon_SDK 3.5.1, HEXAGON_Tools 8.3.07. [Master branch](https://github.com/halide/Halide/commit/6c255c6fd1e2fed7c106e3812e7507a886f5b791) of Halide.

The following results in a sysmalloc Assertion error, then Aborts (core dumped). 

> halide_remote_load_library(libhalide_shared_runtime.so) -> test: malloc.c:2379: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.

Any idea what might be wrong?

```bash
$ cd apps/blur
$ export HL_HEXAGON_SIM_REMOTE=../../src/runtime/hexagon_remote/bin/v62/hexagon_sim_remote
$ export HL_HEXAGON_TOOLS=/home/user/Qualcomm/Hexagon_SDK/3.5.1/tools/HEXAGON_Tools/8.3.07/Tools
$ LD_LIBRARY_PATH=../../src/runtime/hexagon_remote/bin/host/:$HL_HEXAGON_TOOLS/lib/iss/:. HL_TARGET=host-hvx_128-debug make test

g++ -O3 -std=c++11 -I /home/user/Development/halide_git/distrib/include/ -I /home/user/Development/halide_git/distrib/tools/  -Wall -Werror -Wno-unused-function -Wcast-qual -Wignored-qualifiers -Wno-comment -Wsign-compare -Wno-unknown-warning-option -Wno-psabi halide_blur_generator.cpp /home/user/Development/halide_git/distrib/tools/GenGen.cpp -o bin/host/halide_blur.generator -L /home/user/Development/halide_git/distrib/lib/ -lHalide  -ldl -lpthread -lz -ltinfo
bin/host/halide_blur.generator -g halide_blur -e static_library,h,registration,stmt,assembly -o bin/host-hvx_128-debug target=host-hvx_128-debug
g++ -O3 -std=c++11 -I /home/user/Development/halide_git/distrib/include/ -I /home/user/Development/halide_git/distrib/tools/  -Wall -Werror -Wno-unused-function -Wcast-qual -Wignored-qualifiers -Wno-comment -Wsign-compare -Wno-unknown-warning-option -Wno-psabi -fopenmp -Wall -O2 -Ibin/host-hvx_128-debug test.cpp bin/host-hvx_128-debug/halide_blur.a -o bin/host-hvx_128-debug/test  -ldl -lpthread -lz -ltinfo
bin/host-hvx_128-debug/test
Entering Pipeline halide_blur
Target: x86-64-linux-avx-avx2-debug-f16c-fma-hvx_128-sse41
 Input Buffer input: buffer(0, 0x0, 0x7fd4062ba080, 1, uint16, {0, 648, 1}, {0, 482, 648})
 Output Buffer blur_y: buffer(0, 0x0, 0x560b8fd27700, 0, uint16, {0, 640, 1}, {0, 480, 640})
Hexagon: init_hexagon_runtime (user_context: 0x0)
    halide_get_library_symbol('halide_hexagon_remote_load_library') -> 
        0x7fd40692fab7
    halide_get_library_symbol('halide_hexagon_remote_get_symbol_v4') -> 
        0x7fd40692fce2
    halide_get_library_symbol('halide_hexagon_remote_run') -> 
        0x7fd40692fe7a
    halide_get_library_symbol('halide_hexagon_remote_release_library') -> 
        0x7fd4069303b1
    halide_get_library_symbol('halide_hexagon_host_malloc_init') -> 
        0x7fd406930560
    halide_get_library_symbol('halide_hexagon_host_malloc_deinit') -> 
        0x7fd406930567
    halide_get_library_symbol('halide_hexagon_host_malloc') -> 
        0x7fd40693056e
    halide_get_library_symbol('halide_hexagon_host_free') -> 
        0x7fd4069305cf
    halide_get_library_symbol('halide_hexagon_remote_poll_log') -> 
        0x0
    halide_get_library_symbol('halide_hexagon_remote_poll_profiler_state') -> 
        0x7fd4069305f1
    halide_get_library_symbol('halide_hexagon_remote_profiler_set_current_func') -> 
        0x7fd40693061d
    halide_get_library_symbol('halide_hexagon_remote_power_hvx_on') -> 
        0x0
    halide_get_library_symbol('halide_hexagon_remote_power_hvx_off') -> 
        0x0
    halide_get_library_symbol('halide_hexagon_remote_set_performance') -> 
        0x0
    halide_get_library_symbol('halide_hexagon_remote_set_performance_mode') -> 
        0x0
    halide_get_library_symbol('halide_hexagon_remote_set_thread_priority') -> 
        0x0
Hexagon: halide_hexagon_initialize_kernels (user_context: 0x0, state_ptr: 0x560b8e25b3c0, *state_ptr: 0x0, code: 0x560b8e241660, code_size: 8752)
, code: 0x560b8e2438a0, code_size: 57944)
    Initializing shared runtime
    Writing shared object 'libhalide_shared_runtime.so'
    halide_remote_load_library(libhalide_shared_runtime.so) -> test: malloc.c:2379: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
make: *** [Makefile:33: test] Aborted (core dumped)
rm bin/host-hvx_128-debug/halide_blur.a

```

**Comment by andrewjong** [comment]
By the way, if I'm specifically interested in Halide for Qualcomm's Hexagon DSP, should I be working with the Halide distribution from the Qualcomm Developer Network? Or should I work with the code in this repo?

**Comment by dpalermo** [comment]
The assertion you are seeing is coming from the libc malloc on your host system:

    malloc.c:2379: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.

This is a known problem with the current HEXAGON_Tools (specifically how the HexagonWrapper class is declared/allocated in hexagon_sim_remote & libwrapper.so when each are built using different toolsets) which has been recently fixed and will be available in the next SDK release (coming later this year).  To get around this issue, instead of using the hexagon_sim_remote from the Halide github (which is most likely compiled with a different toolset), you can either
    - use the hexagon_sim_remote from the SDK in 3.5.1/tools/HALIDE_Tools/2.2.03/Halide/bin/hexagon_sim_remote.
    - or switch to an older SDK/toolset to match what the hexagon_sim_remote from github was built with...8.1.05...which is available in SDK 3.3.2:
        $ strings hexagon_sim_remote | grep LLVM
        QuIC LLVM Hexagon Clang version 8.1.05 (based on LLVM 4.0.0)

Also, regarding your second question, if you are developing Halide apps for Hexagon (as opposed to working on the Halide compiler itself), you are probably better off using the Halide provided in the SDK distribution (see tools/HALIDE_Tools/2.20.3/Halide) as its runtime was built with the matching Hexagon tools.   There are also Hexagon specific examples available in tools/HALIDE_Tools/2.2.03/Halide/Examples you might find helpful.

**Comment by andrewjong** [comment]
Thanks much!
